agent/connect/ca/provider_vault.go

```diff
@@ -174,9 +178,18 @@ func (v *VaultProvider) generateIntermediateCSR() (string, error) {
 			"require_cn":       false,
 		})
 		if err != nil {
-			return "", err
+			return err
 		}
 	}
+	v.setupIntermediatePKIPathDone = true
```

Comment:
Does this need a mutex or is it only accessed by a single goroutine?

Reply:
Good question! There are two different flows. 

The leader in the primary dc calls this the first time in `initializeRootCA` while holding `caProviderReconfigurationLock`: https://github.com/hashicorp/consul/blob/ad7fb3ab95cfcf39c2ed8cd1a3634cc79820aa76/agent/consul/leader_connect.go#L262-L266

The leader in a secondary dc calls this the first time in `initializeSecondaryCA` while holding `caProviderReconfigurationLock`: https://github.com/hashicorp/consul/blob/ad7fb3ab95cfcf39c2ed8cd1a3634cc79820aa76/agent/consul/leader_connect.go#L346-L349

An error will prevent the initialization to move and and it will be retried later. I think this is safe to do. However I considered using `atomic` just to make sure we don't shoot our foot in case the other code changes.

Reply:
Sounds good.